pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,492 @@
+	defer t.clearConnectionInProgress(node)
+
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v is connecting to node %v", t.ownID, node))
+
+	addr, found := t.getNodeAddr(node)
+	if !found {
+		t.logger.Log(logging.LevelError, "failed to get node address", "src", t.ownID, "dst", node)
+		return
+	}
+	info, err := peer.AddrInfoFromP2pAddr(addr)
+	if err != nil {
+		t.logger.Log(logging.LevelError, "failed to parse addr", "addr", addr, "err", err)
+		return
+	}
+
+	t.host.Peerstore().AddAddrs(info.ID, info.Addrs, PermanentAddrTTL)
+
+	s, err := t.openStream(ctx, info.ID)
+	if err != nil {
+		t.logger.Log(logging.LevelError, "failed to open stream to node", "addr", addr, "node", node, "err", err)
+		return
+	}
+
+	t.addOutboundStream(node, s)
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s has connected to node %s", t.ownID.Pb(), node.Pb()))
+}
+
+func (t *Transport) openStream(ctx context.Context, dest peer.ID) (network.Stream, error) {
+	//  We need the simplest retry mechanism due to the fact that the underlying libp2p's NewStream function dials once:
+	// https://github.com/libp2p/go-libp2p/blob/7828f3e0797e0a7b7033fa5e8be9b94f57a4c173/p2p/net/swarm/swarm.go#L358
+	t.logger.Log(logging.LevelDebug, "opening stream to peer", "src", t.ownID, "dst", dest)
+
+	var streamErr error
+	for i := 0; i < retryAttempts; i++ {
+		sctx, cancel := context.WithTimeout(ctx, maxConnectingTimeout)
+
+		s, streamErr := t.host.NewStream(sctx, dest, ProtocolID)
+		cancel()
+
+		if streamErr == nil {
+			return s, nil
+		}
+
+		if i >= noLoggingErrorAttempts {
+			t.logger.Log(
+				logging.LevelError, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, streamErr))
+		} else {
+			t.logger.Log(
+				logging.LevelInfo, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, streamErr))
+		}
+
+		delay := time.NewTimer(retryTimeout)
+		select {
+		case <-delay.C:
+			continue
+		case <-ctx.Done():
+			if !delay.Stop() {
+				<-delay.C
+			}
+			return nil, fmt.Errorf("%s opening stream to %s: context closed", t.ownID, dest)
+		}
+	}
+	return nil, fmt.Errorf("%s failed to open stream to %s: %w", t.ownID, dest, streamErr)
+}
+
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, payload *messagepb.Message) error {
+	var nodeErr *nodeInfoError
+	var streamErr *streamInfoError
+	s, err := t.getNodeStream(dest)
+	if errors.As(err, &nodeErr) {
+		return err
+	}
+	if errors.As(err, &streamErr) {
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, dest, t.connWg)
+		return err
+	}
+
+	outBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	msg := TransportMessage{t.ownID.Pb(), outBytes}
+	buf := new(bytes.Buffer)
+	err = msg.MarshalCBOR(buf)
+	if err != nil {
+		return fmt.Errorf("failed to CBOR marshal message: %w", err)
+	}
+
+	if ctx.Err() != nil {
+		return err
+	}
+
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		w.Flush()
```

Comment:
IIUC, `Flush` may write the buffered data only partially and return `io.ErrShortWrite` ([source](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/bufio/bufio.go;drc=99f1bf54eb502e123e60c605212a616146fbe86a;l=631)). Have you checked if libp2p stream writer may write less bytes than requested?

Reply:
Yes, good catch. That is a debt from the initial implementation. Fixed

Reply:
I wonder if libp2p itself can write the data partially and return `io.ErrShortWrite` without the stream being broken

Reply:
Can a short write happen even without the connection breaking? If that's the case, we might need to forcefully break the connection and create a new one (or somehow else tell the other end that whatever comes next is a new message, not the continuation of the partially sent one)...

Reply:
Created an issue to address this - https://github.com/filecoin-project/mir/issues/255

Reply:
I don't know IIU the issue correctly, but I honestly don't think this is possible. Libp2p expects the buffer to be explicitly flushed so it can notify the length of the stream being sent to other side (all stream communications are prefixed by a `VarInt` notifying the receiver the amount of bytes to read from the buffer, and buffers are reused on top of the existing connections).